controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -686,14 +665,24 @@ public String retrieveDelegationToken() {
     }
 
     @Override
+    @Synchronized
     public void close() throws Exception {
         timeoutService.stopAsync();
         timeoutService.awaitTerminated();
-        if (commitEventEventStreamWriter != null) {
-            commitEventEventStreamWriter.close();
+        CompletableFuture<Void> commitCloseFuture = commitWriterFuture.thenAccept(EventStreamWriter::close);
+        CompletableFuture<Void> abortCloseFuture = abortWriterFuture.thenAccept(EventStreamWriter::close);
+
+        // since we do the checks under synchronized block, we know the promise for writers cannot be completed 
+        // by anyone if we have the lock and we can simply cancel the futures.  
+        if (commitWriterFuture.isDone()) {
```

Comment:
Wouldn't it be better if you wait a bit for these guys to finish? perhaps have a shutdown timeout of a few seconds, then invoke this with `get(Timeout)` so you don't have to do this complicated thing here.

Reply:
Close is synchronized. So for the production code path, either initialized or close is called. 
if close is called before initialized, then writers are never created. 
if initialize is called before close, writers are initialized and futures are complete. 
So we dont have to wait anyway.